File: code/packages/typescript/transistors/tests/ttl_gates.test.ts
/**
 * Tests for TTL logic gates (historical BJT-based).
 */

import { describe, it, expect } from "vitest";
import { TTLNand, RTLInverter } from "../src/ttl_gates.js";

describe("TTLNand", () => {
  it("should implement NAND truth table", () => {
    const nand = new TTLNand();
    expect(nand.evaluateDigital(0, 0)).toBe(1);
    expect(nand.evaluateDigital(0, 1)).toBe(1);
    expect(nand.evaluateDigital(1, 0)).toBe(1);
    expect(nand.evaluateDigital(1, 1)).toBe(0);
  });

  it("should dissipate milliwatts of static power", () => {
    const nand = new TTLNand();
    expect(nand.staticPower).toBeGreaterThan(1e-3);
  });

  it("output LOW should be near Vce_sat", () => {
    const nand = new TTLNand();
    const result = nand.evaluate(5.0, 5.0);
    expect(result.voltage).toBeLessThan(0.5);
    expect(result.logicValue).toBe(0);
  });

  it("output HIGH should be near Vcc - 0.7V", () => {
    const nand = new TTLNand();
    const result = nand.evaluate(0.0, 0.0);
    expect(result.voltage).toBeGreaterThan(3.0);
    expect(result.logicValue).toBe(1);
  });

  it("propagation delay should be in nanosecond range", () => {
    const nand = new TTLNand();
    const result = nand.evaluate(5.0, 5.0);
    expect(result.propagationDelay).toBeGreaterThan(1e-9);
    expect(result.propagationDelay).toBeLessThan(100e-9);
  });

  it("should reject invalid digital input", () => {
    const nand = new TTLNand();
    expect(() => nand.evaluateDigital(2, 0)).toThrow();
  });

  it("should respect custom Vcc", () => {
    const nand = new TTLNand(3.3);
    expect(nand.vcc).toBe(3.3);
  });
});

describe("RTLInverter", () => {
  it("should implement NOT truth table", () => {
    const inv = new RTLInverter();
    expect(inv.evaluateDigital(0)).toBe(1);
    expect(inv.evaluateDigital(1)).toBe(0);
  });

  it("input LOW should produce output near Vcc", () => {
    const inv = new RTLInverter();
    const result = inv.evaluate(0.0);
    expect(result.voltage).toBeGreaterThan(4.0);
    expect(result.logicValue).toBe(1);
  });

  it("input HIGH should produce output near GND", () => {
    const inv = new RTLInverter();
    const result = inv.evaluate(5.0);
    expect(result.voltage).toBeLessThan(1.0);
    expect(result.logicValue).toBe(0);
  });

  it("propagation delay should be slower than TTL", () => {
    const inv = new RTLInverter();
    const result = inv.evaluate(5.0);
    expect(result.propagationDelay).toBeGreaterThan(10e-9);
  });

  it("should reject invalid digital input", () => {
    const inv = new RTLInverter();
    expect(() => inv.evaluateDigital(true as unknown as number)).toThrow();
  });

  it("should respect custom resistor values", () => {
    const inv = new RTLInverter(5.0, 5000, 2000);
    expect(inv.rBase).toBe(5000);
    expect(inv.rCollector).toBe(2000);
  });
});

File: code/packages/rust/transistors/tests/ttl_gates_test.rs
//! Tests for TTL logic gates (historical BJT-based).

use transistors::ttl_gates::{RTLInverter, TTLNand};

// =========================================================================
// TTL NAND Tests
// =========================================================================

#[test]
fn ttl_nand_truth_table() {
    let nand = TTLNand::new(None, None);
    assert_eq!(nand.evaluate_digital(0, 0).unwrap(), 1);
    assert_eq!(nand.evaluate_digital(0, 1).unwrap(), 1);
    assert_eq!(nand.evaluate_digital(1, 0).unwrap(), 1);
    assert_eq!(nand.evaluate_digital(1, 1).unwrap(), 0);
}

#[test]
fn ttl_nand_static_power_milliwatts() {
    // TTL gates dissipate milliwatts even when idle
    let nand = TTLNand::new(None, None);
    assert!(nand.static_power() > 1e-3);
}

#[test]
fn ttl_nand_output_voltage_low() {
    // Output LOW should be near Vce_sat (~0.2V)
    let nand = TTLNand::new(None, None);
    let result = nand.evaluate(5.0, 5.0);
    assert!(result.voltage < 0.5);
    assert_eq!(result.logic_value, 0);
}

#[test]
fn ttl_nand_output_voltage_high() {
    // Output HIGH should be near Vcc - 0.7V
    let nand = TTLNand::new(None, None);
    let result = nand.evaluate(0.0, 0.0);
    assert!(result.voltage > 3.0);
    assert_eq!(result.logic_value, 1);
}

#[test]
fn ttl_nand_propagation_delay() {
    // TTL should have propagation delay in nanosecond range
    let nand = TTLNand::new(None, None);
    let result = nand.evaluate(5.0, 5.0);
    assert!(result.propagation_delay > 1e-9);
    assert!(result.propagation_delay < 100e-9);
}

#[test]
fn ttl_nand_rejects_invalid_input() {
    let nand = TTLNand::new(None, None);
    assert!(nand.evaluate_digital(2, 0).is_err());
}

#[test]
fn ttl_nand_custom_vcc() {
    // Custom Vcc should be respected
    let nand = TTLNand::new(Some(3.3), None);
    assert_eq!(nand.vcc, 3.3);
}

// =========================================================================
// RTL Inverter Tests
// =========================================================================

#[test]
fn rtl_inverter_truth_table() {
    let inv = RTLInverter::new(None, None, None, None);
    assert_eq!(inv.evaluate_digital(0).unwrap(), 1);
    assert_eq!(inv.evaluate_digital(1).unwrap(), 0);
}

#[test]
fn rtl_inverter_output_voltage_high() {
    // Input LOW -> output near Vcc
    let inv = RTLInverter::new(None, None, None, None);
    let result = inv.evaluate(0.0);
    assert!(result.voltage > 4.0);
    assert_eq!(result.logic_value, 1);
}

#[test]
fn rtl_inverter_output_voltage_low() {
    // Input HIGH -> output near GND
    let inv = RTLInverter::new(None, None, None, None);
    let result = inv.evaluate(5.0);
    assert!(result.voltage < 1.0);
    assert_eq!(result.logic_value, 0);
}

#[test]
fn rtl_inverter_propagation_delay() {
    // RTL should be slower than TTL
    let inv = RTLInverter::new(None, None, None, None);
    let result = inv.evaluate(5.0);
    assert!(result.propagation_delay > 10e-9);
}

#[test]
fn rtl_inverter_rejects_invalid_input() {
    let inv = RTLInverter::new(None, None, None, None);
    assert!(inv.evaluate_digital(2).is_err());
}

#[test]
fn rtl_inverter_custom_resistors() {
    // Custom resistor values should be respected
    let inv = RTLInverter::new(None, Some(5000.0), Some(2000.0), None);
    assert_eq!(inv.r_base, 5000.0);
    assert_eq!(inv.r_collector, 2000.0);
}
